### PR TITLE
APEXCORE-504 - Possible race condition in StreamingContainerAgent.getStreamCodec()

### PR DESCRIPTION
**Pull request by vrozov**
@PramodSSImmaneni or @tweise Please review

### CODE REVIEW

File: engine/src/main/java/com/datatorrent/stram/StreamingContainerAgent.java [hunk 1]
```diff
@@ -215,11 +219,11 @@ public void addOperatorRequest(StramToNodeRequest r)
           for (PTOperator.PTInput input : out.sinks) {
             // Create mappings for all non-inline operators
             if (input.target.getContainer() != out.source.getContainer()) {
-              InputPortMeta inputPortMeta = getIdentifyingInputPortMeta(input);
-              StreamCodec<?> streamCodecInfo = getStreamCodec(inputPortMeta);
-              Integer id = physicalPlan.getStreamCodecIdentifier(streamCodecInfo);
+              final StreamCodec<?> streamCodec = getIdentifyingInputPortMeta(input).getStreamCodec();
+              final Integer id = physicalPlan.getStreamCodecIdentifier(streamCodec);
```
Comment (pramodin) [inline engine/src/main/java/com/datatorrent/stram/StreamingContainerAgent.java:223]: Why the addition of finals in these two lines, does it specify anything significant?

File: engine/src/main/java/com/datatorrent/stram/plan/logical/LogicalPlan.java [hunk 2]
```diff
@@ -309,6 +311,20 @@ public void sendMetrics(Collection<String> metricNames)
...
 
+    public StreamCodec<?> getStreamCodec()
+    {
+      return attributes.get(STREAM_CODEC);
+    }
+
+    void setStreamCodec(StreamCodec<?> streamCodec)
+    {
+      if (streamCodec != null) {
+        StreamCodec<?> oldStreamCodec = attributes.put(STREAM_CODEC, streamCodec);
```
Comment (pramodin) [inline engine/src/main/java/com/datatorrent/stram/plan/logical/LogicalPlan.java:322]: I know we did this in the branch but I don't think it is right to add this back to the attribute list as it changes the attributes from what the user originally specified. How about having the getStreamCodec do the work like it was doing before.

File: engine/src/main/java/com/datatorrent/stram/StreamingContainerAgent.java (hunk 1 above)
Comment (vrozov) [reply engine/src/main/java/com/datatorrent/stram/StreamingContainerAgent.java:223]: I refactored the code using IntelliJ and it adds final by default.

File: engine/src/main/java/com/datatorrent/stram/plan/logical/LogicalPlan.java (hunk 2 above)
Comment (vrozov) [reply engine/src/main/java/com/datatorrent/stram/plan/logical/LogicalPlan.java:322]: Reverting to the original implementation of getStreamCodec() will introduce the original problem back.

File: engine/src/main/java/com/datatorrent/stram/StreamingContainerAgent.java (hunk 1 above)
Comment (pramodin) [reply engine/src/main/java/com/datatorrent/stram/StreamingContainerAgent.java:223]: If it is not needed can we remove the final

File: engine/src/main/java/com/datatorrent/stram/plan/logical/LogicalPlan.java (hunk 2 above)
Comment (pramodin) [reply engine/src/main/java/com/datatorrent/stram/plan/logical/LogicalPlan.java:322]: Can you jog my memory. I thought the problem was that the attribute map is not thread safe. Was there another problem as well?

File: engine/src/main/java/com/datatorrent/stram/StreamingContainerAgent.java (hunk 1 above)
Comment (vrozov) [reply engine/src/main/java/com/datatorrent/stram/StreamingContainerAgent.java:223]: Why is it necessary to remove final? The variables do not need to be changed.

File: engine/src/main/java/com/datatorrent/stram/plan/logical/LogicalPlan.java (hunk 2 above)
Comment (vrozov) [reply engine/src/main/java/com/datatorrent/stram/plan/logical/LogicalPlan.java:322]: Correct, the original implementation of getStreamCodec() is not thread-safe.

**Pull request merged**